### ISSUE TITLE
the question about the positive target number

### ISSUE THREAD

**Issue by lihao930104** [tianzhi0549/FCOS#110]
hi, i find that with five feature maps which have 112 56 28 14 7 side lenth respectIvely just get positive target with number about 10 - 500 ,but the number of pixels of the feature maps are 16709 in my model, is it normal? i want to know how many positive target do you get with your algorithm.thanks

**Comment by tianzhi0549** [comment]
@lihao930104 I suggest you just go ahead to see what will happen. It is difficult to see how many positive samples are normal.

**Comment by lihao930104** [comment]
oh,i just forget to say that i rewrite your algorithm by tensorflow .the classification focal loss descend quickly to near to zero, but all the class prediction to the test dataset is all near to zero likely. i wonder that it is because the imbalance of the positive and negative sample .so the loss = -(t == class_range).float() * term1 * alpha - ((t != class_range) * (t >= 0)).float() * term2 * (1 - alpha)
is dominated by the negative sample.i think maybe my target assign algorithm hay some bug.but i want to confirm this problem first

**Comment by tianzhi0549** [comment]
@lihao930104 I am not sure the exact number of positive samples. I suggest you run the pytorch code and `print(len(pos_inds))` to see the number. The code is at
https://github.com/tianzhi0549/FCOS/blob/f0a9731dac1346788cc30d5751177f2695caaa1f/fcos_core/modeling/rpn/fcos/loss.py#L162

**Comment by lihao930104** [comment]
i see,thanks

**Comment by phy12321** [comment]
@lihao931104
Hi, I'm rewriting the FCOS with mxnet and come up with the same problem, I guess it is the imbalance between positive target and negative target that leads the all class prediction to zeros.However ,I didn't find the solution in source code ,seems there is no such problem in the origin work, I'm really confused. Have you fixed it ? Thanks!

**Comment by phy12321** [comment]
@lihao930104
Hi, I'm rewriting the FCOS with mxnet and come up with the same problem, I guess it is the imbalance between positive target and negative target that leads the all class prediction to zeros.However ,I didn't find the solution in source code ,seems there is no such problem in the origin work, I'm really confused. Have you fixed it ? Thanks!

**Comment by phy12321** [comment]
@lihao930104
Hi, I'm rewriting the FCOS with mxnet and come up with the same problem, I guess it is the imbalance between positive target and negative target that leads the all class prediction to zeros.However ,I didn't find the solution in source code ,seems there is no such problem in the original work, I'm really confused. Have you fixed it ? Thanks!

**Comment by tianzhi0549** [comment]
@phy12321 are you using focal loss?

**Comment by phy12321** [comment]
@tianzhi0549 
yes, I am using sigmoid focal loss

**Comment by tianzhi0549** [comment]
@lihao930104 Did you initialize the bias as in https://github.com/tianzhi0549/FCOS/blob/efb76e48e6490a93cc8b6b5dc93738fa1df34af5/fcos_core/modeling/rpn/fcos/fcos.py#L75?

**Comment by phy12321** [comment]
@lihao930104
i dont think it is the reason that cause all zeros predicion. Any adivice? I wonder how you fix this problem? Thanks!

**Comment by tianzhi0549** [comment]
@phy12321 This initialization is crucial to focal loss. Please refer to the paper of focal loss for details.

**Comment by lihao930104** [comment]
I still have not fix the problem,if you trained the model with the pre-trained weights, prediction scores will located at the interval which seems normal(but all scores are also similar so I  doubt that the model is not trained correctly ).

**Comment by phy12321** [comment]
I see, but loading pre-trained weights doesn't work for me. Maybe there is some bugs in my network. thanks!

**Comment by lichenhao608** [comment]
I also rewrite code by TensorFlow and follow exactly the original code is. However, it yields that the focal loss is extremely high. I don't know if it is the problem of the code or it will drop after epochs.

**Comment by tianzhi0549** [comment]
@lihao930104 @phy12321 this project (https://github.com/hux999/tf-fcos) might be helpful.

**Comment by lihao930104** [comment]
@tianzhi0549 thanks!
